File: py/src/apptheory/naming.py
from __future__ import annotations

import re

_non_alnum = re.compile(r"[^a-z0-9-]+")
_multi_dash = re.compile(r"-+")


def _sanitize_part(value: str) -> str:
    out = str(value or "").strip().lower()
    if not out:
        return ""
    out = out.replace("_", "-").replace(" ", "-")
    out = _non_alnum.sub("-", out)
    out = _multi_dash.sub("-", out)
    return out.strip("-")


def normalize_stage(stage: str) -> str:
    value = str(stage or "").strip().lower()
    if value in {"prod", "production", "live"}:
        return "live"
    if value in {"dev", "development"}:
        return "dev"
    if value in {"stg", "stage", "staging"}:
        return "stage"
    if value in {"test", "testing"}:
        return "test"
    if value == "local":
        return "local"
    return _sanitize_part(value)


def base_name(app_name: str, stage: str, tenant: str = "") -> str:
    app = _sanitize_part(app_name)
    ten = _sanitize_part(tenant)
    stg = normalize_stage(stage)
    if ten:
        return f"{app}-{ten}-{stg}"
    return f"{app}-{stg}"


def resource_name(app_name: str, resource: str, stage: str, tenant: str = "") -> str:
    app = _sanitize_part(app_name)
    ten = _sanitize_part(tenant)
    res = _sanitize_part(resource)
    stg = normalize_stage(stage)
    if ten:
        return f"{app}-{ten}-{res}-{stg}"
    return f"{app}-{res}-{stg}"

File: pkg/naming/naming.go
package naming

import (
	"regexp"
	"strings"
)

var (
	nonAlnum  = regexp.MustCompile(`[^a-z0-9-]+`)
	multiDash = regexp.MustCompile(`-+`)
)

func sanitizePart(value string) string {
	value = strings.ToLower(strings.TrimSpace(value))
	if value == "" {
		return ""
	}
	value = strings.ReplaceAll(value, "_", "-")
	value = strings.ReplaceAll(value, " ", "-")
	value = nonAlnum.ReplaceAllString(value, "-")
	value = multiDash.ReplaceAllString(value, "-")
	value = strings.Trim(value, "-")
	return value
}

// NormalizeStage maps stage aliases to canonical values.
//
// Canonical stages are lowercased and safe for typical resource naming schemes.
func NormalizeStage(stage string) string {
	stage = strings.ToLower(strings.TrimSpace(stage))
	switch stage {
	case "prod", "production", "live":
		return "live"
	case "dev", "development":
		return "dev"
	case "stg", "stage", "staging":
		return "stage"
	case "test", "testing":
		return "test"
	case "local":
		return "local"
	default:
		return sanitizePart(stage)
	}
}

// BaseName returns a deterministic base name:
// - <app>-<stage>
// - <app>-<tenant>-<stage> (when tenant is provided)
func BaseName(appName, stage, tenant string) string {
	app := sanitizePart(appName)
	tenant = sanitizePart(tenant)
	stage = NormalizeStage(stage)

	parts := []string{app}
	if tenant != "" {
		parts = append(parts, tenant)
	}
	if stage != "" {
		parts = append(parts, stage)
	}
	return strings.Join(parts, "-")
}

// ResourceName returns a deterministic resource name:
// - <app>-<resource>-<stage>
// - <app>-<tenant>-<resource>-<stage> (when tenant is provided)
func ResourceName(appName, resource, stage, tenant string) string {
	app := sanitizePart(appName)
	tenant = sanitizePart(tenant)
	resource = sanitizePart(resource)
	stage = NormalizeStage(stage)

	parts := []string{app}
	if tenant != "" {
		parts = append(parts, tenant)
	}
	if resource != "" {
		parts = append(parts, resource)
	}
	if stage != "" {
		parts = append(parts, stage)
	}
	return strings.Join(parts, "-")
}
